File: post/src/main/java/com/example/learningwebflux/post/integrationeventreceivers/usersignedup/CreateUserIEReceiver.java
package com.example.learningwebflux.post.integrationeventreceivers.usersignedup;

import com.example.learningwebflux.common.integrationeventbus.IntegrationEventReceiver;
import com.example.learningwebflux.common.integrationevents.UserSignedUpIE;
import com.example.learningwebflux.post.user.User;
import com.example.learningwebflux.post.user.repository.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
@AllArgsConstructor
public class CreateUserIEReceiver implements IntegrationEventReceiver<UserSignedUpIE> {
    @Override
    public Mono<Void> receive(UserSignedUpIE event) {
        var user = new User(event.id());

        return userRepository.save(user);
    }


    private final UserRepository userRepository;
}

File: postanalytics/src/main/kotlin/com/example/learningwebflux/postanalytics/iereceivers/usersignedup/CreateUserIEReceiver.kt
package com.example.learningwebflux.postanalytics.iereceivers.usersignedup

import com.example.learningwebflux.common.integrationeventbus.IntegrationEventReceiver
import com.example.learningwebflux.common.integrationevents.UserSignedUpIE
import com.example.learningwebflux.postanalytics.user.User
import com.example.learningwebflux.postanalytics.user.UserRepository
import kotlinx.coroutines.reactor.mono
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
internal class CreateUserIEReceiver(private val userRepository: UserRepository) : IntegrationEventReceiver<UserSignedUpIE> {
    override fun receive(event: UserSignedUpIE): Mono<Void> {
        return mono { userRepository.save(User(event.id)) }
            .then()
    }
}